File: web/backend-go/runner.go
package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

var (
	gameBin    = envOr("GAME_BIN", "/app/game/Game")
	configFile = envOr("CONFIG_FILE", "data/default.cnf")

	fallbackAIs = []string{"Demo", "JohnnyBGood"}
)

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func gameAvailable() bool {
	info, err := os.Stat(gameBin)
	return err == nil && !info.IsDir()
}

func listAIs() []string {
	if !gameAvailable() {
		return nil
	}
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	out, err := exec.CommandContext(ctx, gameBin, "--list").Output()
	if err != nil {
		return fallbackAIs
	}

	var ais []string
	for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
		if s := strings.TrimSpace(line); s != "" {
			ais = append(ais, s)
		}
	}
	if len(ais) == 0 {
		return fallbackAIs
	}
	return ais
}

func runGame(seed int, players []string) (string, error) {
	if !gameAvailable() {
		return "", fmt.Errorf("game binary not found — C++ engine not compiled")
	}

	if len(players) == 0 {
		ais := listAIs()
		numPlayers := 4
		players = make([]string, numPlayers)
		for i := range players {
			players[i] = ais[i%len(ais)]
		}
	}

	args := []string{"-i", configFile, "--seed", strconv.Itoa(seed)}
	args = append(args, players...)

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, gameBin, args...)
	out, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			stderr := string(exitErr.Stderr)
			if len(stderr) > 500 {
				stderr = stderr[:500]
			}
			return "", fmt.Errorf("game exited with code %d: %s", exitErr.ExitCode(), stderr)
		}
		return "", err
	}
	return string(out), nil
}

File: web/backend/runner.py
"""Run the C++ game engine and return the raw replay text."""
from __future__ import annotations

import subprocess
from pathlib import Path

GAME_BIN = Path("/app/game/Game")
CONFIG_FILE = Path(__file__).parent / "data" / "default.cnf"

AVAILABLE_AIS = ["Demo", "JohnnyBGood"]


def game_available() -> bool:
    return GAME_BIN.exists() and GAME_BIN.is_file()


def list_ais() -> list[str]:
    if not game_available():
        return []
    try:
        r = subprocess.run([str(GAME_BIN), "--list"], capture_output=True, text=True, timeout=5)
        return [line.strip() for line in r.stdout.strip().split("\n") if line.strip()]
    except Exception:
        return AVAILABLE_AIS


def run_game(seed: int, players: list[str] | None = None) -> str:
    if not game_available():
        raise RuntimeError("Game binary not found — C++ engine not compiled")

    if not players:
        ais = list_ais()
        num_players = 4
        players = [ais[i % len(ais)] for i in range(num_players)]

    cmd = [str(GAME_BIN), "-i", str(CONFIG_FILE), "--seed", str(seed)] + players

    result = subprocess.run(
        cmd,
        capture_output=True,
        text=True,
        timeout=30,
    )
    if result.returncode != 0:
        raise RuntimeError(f"Game exited with code {result.returncode}: {result.stderr[:500]}")
    return result.stdout
